### PR TITLE
Skip null checks for positions from blocks that have no nulls in aggr…

### PR DESCRIPTION
**Pull request by radek-starburst**
Using `io.trino.spi.block.Block.mayHaveNull` it can be detected that there are no null positions in the block. Basing on that we skip checking nullability of every postition for such blocks in the aggregation `input` method for performance reason.

Performance results _(cpu time, parquet, unpart, sf1000)_
**tpch: -1.77%** (average for two runs)
**tpcds: -1.61%** (average for two runs)

For queries where aggregation is a significant operator there is more gain, like 4-5% cpu time gain, for example:
_q06,q12_ in _tpch_ and  _q09 / q28_ in _tpcds_
full report:
[serial benchmarks.pdf](https://github.com/trinodb/trino/files/9758095/serial.benchmarks.pdf)

(* ) This is not user-visible or docs only and no release notes are required.
( ) Release notes are required, please propose a release note for me.
( ) Release notes are required, with the following suggested text:
(*) Release note is optional: Improved performance for aggregation

### CODE REVIEW

File: core/trino-main/src/main/java/io/trino/operator/aggregation/AccumulatorCompiler.java [hunk 1]
```diff
@@ -503,12 +503,39 @@ private static BytecodeBlock generateInputForLoop(
         Variable positionVariable = scope.declareVariable(int.class, "position");
         Variable rowsVariable = scope.declareVariable(int.class, "rows");
 
+        // prepare mayHaveNull condition
```
Comment (lukasz-stec) [inline core/trino-main/src/main/java/io/trino/operator/aggregation/AccumulatorCompiler.java:506]: move this block closer to the usage (above `IfStatement mayHaveNullIf`)

File: core/trino-main/src/main/java/io/trino/operator/aggregation/AccumulatorCompiler.java [hunk 2]
```diff
@@ -503,12 +503,39 @@ private static BytecodeBlock generateInputForLoop(
...
         BytecodeBlock block = new BytecodeBlock()
                 .append(arguments)
                 .invokeVirtual(Page.class, "getPositionCount", int.class)
                 .putVariable(rowsVariable)
                 .initializeVariable(positionVariable);
 
+        ForLoop nullCheckLoop = generateInputLoopBody(true, scope, stateField, positionVariable, parameterVariables, lambdaProviderFields, inputFunction, callSiteBinder, grouped, argumentNullable, masksBlock, rowsVariable);
+        ForLoop noNullCheckLoop = generateInputLoopBody(false, scope, stateField, positionVariable, parameterVariables, lambdaProviderFields, inputFunction, callSiteBinder, grouped, argumentNullable, masksBlock, rowsVariable);
+
+        IfStatement mayHaveNullIf = new IfStatement("if(%s)", mayHaveNullCondition).condition(mayHaveNullCondition)
```
Comment (lukasz-stec) [inline core/trino-main/src/main/java/io/trino/operator/aggregation/AccumulatorCompiler.java:523]: Are there tests that cover this case (both edges)?

File: core/trino-main/src/main/java/io/trino/operator/aggregation/AccumulatorCompiler.java [hunk 3]
```diff
@@ -503,12 +503,39 @@ private static BytecodeBlock generateInputForLoop(
...
+                .ifFalse(noNullCheckLoop)
+                .ifTrue(nullCheckLoop);
+
+        block.append(new IfStatement("if(!maskGuaranteedToFilterAllRows(%s, %s))", rowsVariable.getName(), masksBlock.getName())
```
Comment (lukasz-stec) [inline core/trino-main/src/main/java/io/trino/operator/aggregation/AccumulatorCompiler.java:527]: Could you add a comment/documentation on how the generated code loooks like?
I know that this class has "limited" documentation, but I find it very useful when analyzing bytecode generation to see what's the expected result.

**Review by lukasz-stec — APPROVED** [review]
lgtm % comments

File: core/trino-main/src/main/java/io/trino/operator/aggregation/AccumulatorCompiler.java [hunk 4]
```diff
@@ -509,6 +509,33 @@ private static BytecodeBlock generateInputForLoop(
                 .putVariable(rowsVariable)
                 .initializeVariable(positionVariable);
 
+        ForLoop nullCheckLoop = generateInputLoopBody(true, scope, stateField, positionVariable, parameterVariables, lambdaProviderFields, inputFunction, callSiteBinder, grouped, argumentNullable, masksBlock, rowsVariable);
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/operator/aggregation/AccumulatorCompiler.java:512]: commit message is too long

Comment (sopel39) [reply core/trino-main/src/main/java/io/trino/operator/aggregation/AccumulatorCompiler.java:512]: ping

File: core/trino-main/src/main/java/io/trino/operator/aggregation/AccumulatorCompiler.java [hunk 5]
```diff
@@ -509,6 +509,33 @@ private static BytecodeBlock generateInputForLoop(
...
+        ForLoop noNullCheckLoop = generateInputLoopBody(false, scope, stateField, positionVariable, parameterVariables, lambdaProviderFields, inputFunction, callSiteBinder, grouped, argumentNullable, masksBlock, rowsVariable);
+
+        // prepare mayHaveNull condition
+        BytecodeExpression mayHaveNullCondition = BytecodeExpressions.constantFalse();
+        for (int parameterIndex = 0; parameterIndex < parameterVariables.size(); parameterIndex++) {
+            if (!argumentNullable.get(parameterIndex)) {
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/operator/aggregation/AccumulatorCompiler.java:518]: why we need this?

File: core/trino-main/src/main/java/io/trino/operator/aggregation/AccumulatorCompiler.java (hunk 2 above)
Comment (sopel39) [reply core/trino-main/src/main/java/io/trino/operator/aggregation/AccumulatorCompiler.java:523]: Consider adding tests to `TestAccumulatorCompiler`

**Review by sopel39 — COMMENTED** [review]
lgtm % comments

**Review by sopel39 — APPROVED** [review]

**Pull request merged**